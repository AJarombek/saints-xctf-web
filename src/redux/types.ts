/**
 * Types for the Redux store.
 * @author Andrew Jarombek
 * @since 7/23/2020
 */

export type RootState = {
    auth: AuthState,
    logs: LogsState
}

export type LogsState = {
    isFetching: boolean;
    didInvalidate: boolean;
    lastUpdated: number;
    items: Logs;
    feeds: LogFeeds;
    newComments: NewComments;
}

export type AuthState = {
    auth: Auth;
    user: Users;
}

export interface Meta {
    isFetching?: boolean;
    lastUpdated?: number;
    didInvalidate?: boolean;
    serverError?: string;
}

export type Auth = {
    isFetching?: boolean;
    lastUpdated?: number;
    signedInUser?: string;
    status?: string;
}

export type Users = {
    [key: string]: UserMeta;
}

export interface UserMeta extends User, Meta {}

export interface User {
    activation_code?: string;
    class_year?: number;
    deleted?: string;
    description?: string;
    email?: string;
    favorite_event?: string;
    first?: string;
    last?: string;
    last_signin?: string;
    location?: string;
    member_since?: string;
    password?: string;
    salt?: string;
    subscribed?: string;
    username?: string;
    week_start?: string;
}

enum LogType { RUN = "run", BIKE = "bike", SWIM = "swim", OTHER = "other" }
enum Metric { MILES = "miles", KILOMETERS = "kilometers", METERS = "meters" }

export type Log = {
    log_id: number;
    username: string;
    first: string;
    last: string;
    name: string;
    date: string;
    type: LogType;
    distance?: number;
    metric?: Metric;
    miles?: number;
    time?: string;
    pace?: string;
    feel: number;
    location?: string;
    description?: string;
    timeCreated?: string;
    comments?: Comment[];
}

export type Logs = {
    [key: string]: Record<string, Log>;
}

export type LogFeeds = {
    [key: string]: {
        filterBy: string;
        bucket: string;
        pages: LogFeedPages
    };
}

export type LogFeedPages = {
    [key: string]: LogFeedPage
}

export type LogFeedPage = {
    isFetching: boolean;
    lastUpdated: number;
    items: Log[];
    serverError: string;
}

export type Comment = {
    comment_id: number;
    username: string;
    first: string;
    last: string;
    log_id: number;
    time: string;
    content: string;
}

export type NewComments = {
    [key: string]: {
        isFetching: boolean;
        lastUpdated: number;
        created?: boolean;
        serverError?: string;
    }
}
