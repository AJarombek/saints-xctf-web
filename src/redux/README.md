### Overview

Configure Redux as the state management system for the application.

### Files

| Filename                 | Description                                                                |
|--------------------------|----------------------------------------------------------------------------|
| `modules`                | Redux modules containing Reducers, Actions, and Action Creators.           |
| `appState.js`            | Sample application state maintained by Redux.                              |
| `store.js`               | Configuration and creation of the Redux store.                             |
| `types.ts`               | Typescript types used throughout the application.                          |

### Resources

1) [React/Redux Example App #1](https://github.com/erikras/react-redux-universal-hot-example/tree/master/src/redux)
2) [React/Redux Example App #2](https://github.com/goopscoop/ga-react-tutorial/tree/6-reduxActionsAndReducers/src/redux)
3) [React/Redux Example App #3](https://github.com/gothinkster/react-redux-realworld-example-app/tree/master/src)
4) [Redux Ducks Spec](https://github.com/erikras/ducks-modular-redux)
5) [React Router/Redux Setup Steps](https://github.com/supasate/connected-react-router#usage)
6) [Redux Hot Reloading](https://redux.js.org/recipes/configuring-your-store/#hot-reloading)
7) [Redux Thunk](https://github.com/reduxjs/redux-thunk)
8) [Redux/React Official Real World Example](https://github.com/reduxjs/redux/tree/master/examples/real-world)
