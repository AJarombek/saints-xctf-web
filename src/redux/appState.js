/**
 * Sample Redux State for the entire application.
 * @author Andrew Jarombek
 * @since 5/3/2020
 */

const appState = {
  auth: {
    auth: {
      isFetching: false,
      lastUpdated: 1588530309,
      signedIn: true,
      status: "SUCCESS"
    },
    registration: {
      stage: 1,
      first: "Andy",
      last: "Jarombek",
      email: "andrew@jarombek.com",
      username: undefined,
      password: undefined,
      activationCode: undefined,
      teams: {
        saintsXCTFAlumni: {
          status: "accepted",
          groups: {
            alumni: {
              status: "accepted"
            }
          }
        }
      }
    },
    user: {
      andy: {
        isFetching: false,
        didInvalidate: false,
        lastUpdated: 1588530310,
        username: "andy",
        first: "Andrew",
        last: "Jarombek",
        password: "xxxxxx",
        description: "Test Description",
        memberSince: "2016-02-23 12:00:00",
        classYear: 2017,
        location: "New York, NY",
        favoriteEvent: "Cooldown Jog",
        activationCode: "ABCD1234",
        email: "andrew@jarombek.com",
        lastSignIn: "2020-05-03 12:00:00",
        subscribed: "Y",
        logs: []
      }
    },
  },
  profile: {
    tom: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530311
    },
    joe: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530312
    },
    ben: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530313
    }
  },
  group: {
    wmenstf: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530314,
      groupName: "wmenstf",
      groupTitle: "Women's Track & Field",
      description: "",
      weekStart: "monday",
      members: [],
      logs: []
    }
  },
  logs: {
    isFetching: false,
    didInvalidate: false,
    lastUpdated: 1588530315,
    items: {
      1: {
        log_id: 1,
        username: "andy",
        first: "Andrew",
        last: "Jarombek",
        name: "Reservoir Loop",
        location: "New York, NY",
        date: "2020-05-03",
        type: "run",
        distance: 6.37,
        metric: "miles",
        miles: 6.37,
        time: "00:45:37",
        pace: "00:07:10",
        feel: 6,
        description: "",
        timeCreated: "2020-05-03 12:00:00",
        comments: {
          1: {
            commentId: 1,
            username: "andy",
            first: "Andrew",
            last: "Jarombek",
            logId: 1,
            time: "2020-05-03 12:00:00",
            content: "test"
          }
        }
      }
    }
  }
};
