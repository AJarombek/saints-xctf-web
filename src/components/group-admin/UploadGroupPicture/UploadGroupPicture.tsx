/**
 * Component where admins upload a group picture.
 * @author Andrew Jarombek
 * @since 1/28/2021
 */

import React, { useMemo, useState } from 'react';
import UploadPicture from '../../shared/UploadPicture';
import { useDispatch, useSelector } from 'react-redux';
import { GroupMeta, RootState, UploadingGroupPicture } from '../../../redux/types';
import { putGroup, uploadGroupPicture } from '../../../redux/modules/groups';
import { timeout } from '../../../utils/timeout';

interface Props {
  group: GroupMeta;
  groupPictureUrl: string;
}

const UploadGroupPicture: React.FunctionComponent<Props> = ({ group, groupPictureUrl }) => {
  const dispatch = useDispatch();
  const uploadingGroupPictures: Record<string, UploadingGroupPicture> = useSelector(
    (state: RootState) => state.groups.uploadingGroupPicture
  );

  const uploadingGroupPicture: UploadingGroupPicture = useMemo(() => {
    if (group.id) {
      return uploadingGroupPictures[group.id];
    } else {
      return null;
    }
  }, [group.id, uploadingGroupPictures]);

  const [file, setFile] = useState<File>(null);
  const [saving, setSaving] = useState(false);
  const [errorUploading, setErrorUploading] = useState(false);
  const [errorUpdatingGroup, setErrorUpdatingGroup] = useState(false);
  const [uploadSuccess, setUploadSuccess] = useState(false);

  const updateGroup = async (): Promise<void> => {
    const newGroup = { ...group };
    newGroup.grouppic_name = file.name;

    const updatedGroup = await dispatch(putGroup(newGroup));

    if (!updatedGroup) {
      setErrorUpdatingGroup(true);
      setSaving(false);
    } else {
      await timeout(250);

      setFile(null);
      setSaving(false);
      setUploadSuccess(true);
    }
  };

  const onSubmitPicture = async (): Promise<void> => {
    setSaving(true);
    setUploadSuccess(false);
    setErrorUploading(false);
    setErrorUpdatingGroup(false);

    const result = await dispatch(uploadGroupPicture(group.id, file));

    if (result) {
      await updateGroup();
    } else {
      setErrorUploading(true);
      setSaving(false);
    }
  };

  const onCancelPicture = (): void => {
    setFile(null);
  };

  return (
    <UploadPicture
      pictureUrl={groupPictureUrl}
      uploadingPicture={uploadingGroupPicture}
      errorUpdating={errorUpdatingGroup}
      errorUpdatingMessage="Failed to update group details with the new group picture"
      onCloseErrorUpdatingModal={(): void => setErrorUpdatingGroup(false)}
      errorUploading={errorUploading}
      errorUploadingMessage="Failed to upload a new picture for the group"
      onCloseErrorUploadingModal={(): void => setErrorUploading(false)}
      onRetryUpload={onSubmitPicture}
      onRetryUpdate={updateGroup}
      file={file}
      setFile={setFile}
      saving={saving}
      uploadSuccess={uploadSuccess}
      setUploadSuccess={setUploadSuccess}
      onSubmitPicture={onSubmitPicture}
      onCancelPicture={onCancelPicture}
    />
  );
};

export default UploadGroupPicture;
