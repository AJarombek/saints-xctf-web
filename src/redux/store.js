/**
 * Build and configure the redux store, which holds the application state and handles
 * all state updates.
 * @author Andrew Jarombek
 * @since 5/1/2020
 */

import { createBrowserHistory } from 'history';
import { createStore, applyMiddleware } from 'redux';
import { routerMiddleware } from 'connected-react-router';
import createLogger from 'redux-logger';
import reducer from './modules/reducers';

const history = createBrowserHistory();

const configureStore = (data) => {
  const loggerMiddleware = createLogger();
  const middleware = [loggerMiddleware, routerMiddleware(history)];

  const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
  return createStoreWithMiddleware(reducer, data);
};

export { history };
export default configureStore;
