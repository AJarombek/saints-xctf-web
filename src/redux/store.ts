/**
 * Build and configure the redux store, which holds the application state and handles
 * all state updates.
 * @author Andrew Jarombek
 * @since 5/1/2020
 */

import { createBrowserHistory } from 'history';
import { createStore, applyMiddleware, Store, Action } from 'redux';
import { routerMiddleware } from 'connected-react-router';
import { createLogger } from 'redux-logger';
import thunkMiddleware, { ThunkAction } from 'redux-thunk';
import reducer from './modules/reducers';
import { RootState } from './types';

export const history = createBrowserHistory();

export default function configureStore(): Store {
  const loggerMiddleware = createLogger();
  const middleware = [routerMiddleware(history), loggerMiddleware, thunkMiddleware];

  const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
  const store = createStoreWithMiddleware(reducer(history));

  if (process.env.NODE_ENV !== 'production' && module.hot) {
    module.hot.accept('./modules/reducers', () => store.replaceReducer(reducer));
  }

  return store;
}

export type AppThunk<ReturnType = void, State = RootState> = ThunkAction<ReturnType, State, unknown, Action<string>>;
