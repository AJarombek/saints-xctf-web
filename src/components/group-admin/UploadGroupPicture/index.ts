import UploadGroupPicture from './UploadGroupPicture';

export default UploadGroupPicture;
