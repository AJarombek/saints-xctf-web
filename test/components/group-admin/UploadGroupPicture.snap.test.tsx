/**
 * Snapshot test for the UploadGroupPicture component.
 * @author Andrew Jarombek
 * @since 3/5/2021
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { Store } from 'redux';
import { emptyStore } from '../../test-utils/storeMocks';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import UploadGroupPicture from '../../../src/components/group-admin/UploadGroupPicture';
import { alumni } from '../../test-utils/groupMocks';

const mockStore = configureStore([thunk]);

describe('UploadGroupPicture Snapshot Tests', () => {
  let store: Store;

  beforeEach(() => {
    store = mockStore(emptyStore);
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <UploadGroupPicture group={alumni} groupPictureUrl="picture.png" />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
