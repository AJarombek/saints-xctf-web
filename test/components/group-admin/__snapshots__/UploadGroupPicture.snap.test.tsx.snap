// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadGroupPicture Snapshot Tests renders correctly 1`] = `
Array [
  <div
    className="pictureContainer-0-1-1"
  >
    <figure
      className="picture-0-1-2"
    >
      <img
        alt=""
        src="picture.png"
      />
    </figure>
    <div
      className="uploadFile-0-1-9 uploadFileBoxEmpty-0-1-11"
      data-cypress="uploadFile"
      onClick={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
    >
      <input
        accept="image/*"
        className="hiddenInput-0-1-12"
        onChange={[Function]}
        type="file"
      />
      <p
        className="uploadText-0-1-13 uploadTextEmpty-0-1-15"
      >
        Click here or drag and drop a file.
      </p>
    </div>
  </div>,
  <div
    className="note-0-1-8"
  >
    <p>
      NOTE: For best results, crop your image into a square.
    </p>
  </div>,
  <div
    className="actions-0-1-3"
  >
    <div
      className="ajBaseButton-2-2-4 ajBaseButtonDisabled-2-2-6 aj-contained-button ajContainedButton-2-2-1 ajContainedButtonDisabled-2-2-3 submitButton-0-1-4 disabledSubmitButton-0-1-5"
      onClick={null}
    >
      <button
        disabled={true}
        type="button"
      >
        <p>
          Save Picture
        </p>
      </button>
    </div>
    <div
      className="ajBaseButton-2-2-4 ajBaseButtonEnabled-2-2-5 aj-text-button ajTextButton-2-2-7 ajTextButtonEnabled-2-2-8 cancelButton-0-1-7"
      onClick={[Function]}
    >
      <button
        disabled={false}
        type="button"
      >
        Cancel
      </button>
    </div>
  </div>,
]
`;
